compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -7033,27 +7056,30 @@ private STNode parseCompoundBinaryOperator() {
      * @param qualifiers     Preceding top level qualifiers
      * @return Parsed node
      */
-    private STNode parseServiceDeclOrVarDecl(STNode metadata, List<STNode> qualifiers) {
+    private STNode parseServiceDeclOrVarDecl(STNode metadata, STNode publicQualifier, List<STNode> qualifiers) {
         startContext(ParserRuleContext.SERVICE_DECL);
         STNode serviceDeclQualNodeList = extractAndCreateServiceDeclQual(qualifiers);
         STNode serviceKeyword = extractServiceKeyword(qualifiers);
         STNode typeDesc = parseServiceDeclTypeDescriptor(qualifiers);
 
         if (typeDesc != null && typeDesc.kind == SyntaxKind.OBJECT_TYPE_DESC) {
-            return parseServiceDeclOrVarDecl(metadata, serviceDeclQualNodeList, serviceKeyword, typeDesc);
+            return parseServiceDeclOrVarDecl(metadata, publicQualifier, serviceDeclQualNodeList, serviceKeyword,
+                    typeDesc);
         } else {
+            reportInvalidQualifier(publicQualifier);
             return parseServiceDecl(metadata, serviceDeclQualNodeList, serviceKeyword, typeDesc);
         }
     }
 
-    private STNode parseServiceDeclOrVarDecl(STNode metadata, STNode serviceDeclQualNodeList, STNode serviceKeyword,
-                                             STNode typeDesc) {
+    private STNode parseServiceDeclOrVarDecl(STNode metadata, STNode publicQualifier, STNode serviceDeclQualNodeList,
+                                             STNode serviceKeyword, STNode typeDesc) {
         // Reaching here means service keyword is followed by an object type desc.
         // It could either be service declaration or variable declaration with service object type.
         STToken nextToken = peek();
         switch (nextToken.kind) {
             case SLASH_TOKEN:
             case ON_KEYWORD:
+                reportInvalidQualifier(publicQualifier);
```

Comment:
We have a problem here. What's underneath the reportInvalidQualifier() is `addInvalidNodeToNextToken`. Then `public` will be in the leading minutiae of whatever the next token. 

Also, since we have used it in two places (L7082 and L7069 ), maybe we can take the `public` inside the parseServiceDecl() and validate it at once.

Reply:
IMO it's not clear to take `public` as a parameter for `parseServiceDecl`